maxeb markdown spec

fradamt opened the pull request.

michaelneuder committed:
pyspec init commit rebased on f1dff5f67

michaelneuder committed:
update with new spec

michaelneuder committed:
update with new spec

michaelneuder committed:
continue pyspec tests

michaelneuder committed:
constants fixes

michaelneuder committed:
passing lint!

michaelneuder committed:
forks

michaelneuder committed:
activation and exit queue tests

michaelneuder committed:
rename file

djrtwo committed:
add note about not invalidating valid and available blocks

hwwhww committed:
Update other specs accordingly and re-order "Proposer head and reorg helpers" section

djrtwo committed:
Update fork-choice.md

hwwhww committed:
Add `get_proposer_head` and `should_override_forkchoice_update` tests

djrtwo committed:
shift wehre warning is in comments

djrtwo committed:
Merge pull request #3526 from ethereum/validity-note

add note about not invalidating valid and available blocks

hwwhww committed:
Fix `parent_state_advanced` side effect and reorder the conditions

hwwhww committed:
Update config files

hwwhww committed:
Update test format and output `should_override_forkchoice_update` result

nisdas committed:
Update p2p-interface.md

michaelsproul committed:
Clarifications for proposer boost reorgs

michaelsproul committed:
Update test format. Add `get_proposer_head`

hwwhww committed:
Add `validator_is_connected`

hwwhww committed:
Merge branch 'dev' into pr3034

dapplion committed:
Add blob sidecar inclusion proof

dapplion committed:
remove verify_blob_sidecar_signature

dapplion committed:
compute KZG_COMMITMENT_INCLUSION_PROOF_DEPTH

dapplion committed:
List typo

dapplion committed:
doctoc

dapplion committed:
pass lint

dapplion committed:
build tree

dapplion committed:
Update specs/deneb/p2p-interface.md

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

dapplion committed:
fix unit tests

dapplion committed:
doctoc

dapplion committed:
review PR

hwwhww committed:
Move `KZG_COMMITMENT_INCLUSION_PROOF_DEPTH` to preset and cast `int()`

ralexstokes committed:
Update validator.md

Align block production description using the latest APIs with corresponding symbols.

ralexstokes committed:
remove `BLOB_TX_TYPE` from spec (and move to tests)

ralexstokes committed:
Update p2p-interface.md

Suggest readers of `BlobSidecarsByRoot` also verify KZG proofs of blob data.

hwwhww committed:
Add `BLOB_KZG_COMMITMENTS_GINDEX` to "Constant". Use pyspec parser trick to add assertion in pyspec

hwwhww committed:
Fix toc

hwwhww committed:
Fix test

hwwhww committed:
Remove `BLOB_KZG_COMMITMENTS_GINDEX` from the preset files

hwwhww committed:
Fix lint

hwwhww committed:
Add a general `compute_merkle_proof` helper to replace container-specific helpers

dapplion committed:
drop is_valid_merkle_path

dapplion committed:
Update specs/deneb/p2p-interface.md

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

dapplion committed:
drop sidecar alias

dapplion committed:
Update specs/deneb/p2p-interface.md

Co-authored-by: Pop Chunhapanya <haxx.pop@gmail.com>

hwwhww committed:
Enhance `blob_sidecar_inclusion_proof` tests

hwwhww committed:
Fix typing and delete the `signed_sidecar`

djrtwo committed:
Merge pull request #3529 from nisdas/patch-3

Reject Blobs Which Reference Already Finalized Parent Blocks

djrtwo committed:
Merge pull request #3533 from ethereum/ralexstokes-patch-2

`deneb`: Update block production to more precisely reflect the latest flow

hwwhww committed:
Add Merkle proof test

djrtwo committed:
Merge pull request #3534 from ralexstokes/clean-up-deneb-tx-type

remove `BLOB_TX_TYPE` from spec (and move to tests)

dapplion committed:
Add verify_blob_kzg_proof condition

dapplion committed:
Merge branch 'dev' into blob-p2p-proof

djrtwo committed:
Merge pull request #3535 from ethereum/ralexstokes-patch-3

`deneb`: Suggest sidecars by root callers verify blob sidecars

etan-status committed:
Rename `_INDEX` generalized index constants to `_GINDEX`

To assist with readability, use `_GINDEX` suffix when referring to
generalized index constants.

Does neither affect consensus, beacon-API nor builder API.

dapplion committed:
rename to kzg_commitment_inclusion_proof

hwwhww committed:
Remove `BLOB_KZG_COMMITMENTS_GINDEX`

hwwhww committed:
Merge pull request #3034 from michaelsproul/proposer-boost-reorg

Allow honest validators to reorg late blocks

hwwhww committed:
minor refactoring

djrtwo committed:
a few cleanups to sidecar gossip conditions

hwwhww committed:
Bump v1.4.0-beta.4

djrtwo committed:
Merge pull request #3531 from dapplion/blob-p2p-proof

Add blob sidecar inclusion proof

hwwhww committed:
Merge pull request #3538 from ethereum/bump-version

Bump v1.4.0-beta.4

djrtwo committed:
clarify/fix things in blob requests

djrtwo committed:
Merge pull request #3540 from ethereum/blob-by

clarify/fix things in blob requests

etan-status committed:
Remove `SignedBlobSidecar` refs from p2p spec

P2P spec still mentions _signed_ blob sidecars. Remove _signed_ as that
is no longer accurate.

etan-status committed:
Remove unused `DOMAIN_BLOB_SIDECAR`

Blob sidecars are no longer signed, remove signing domain.

etan-status committed:
Update toc

djrtwo committed:
Merge pull request #3542 from etan-status/df-nodomain

Remove unused `DOMAIN_BLOB_SIDECAR`

djrtwo committed:
Merge pull request #3541 from etan-status/df-rmsigned

Remove `SignedBlobSidecar` refs from p2p spec

etan-status committed:
Use fork choice to determine `finalized_checkpoint` in gossip validation

Update gossip validation rules to use the highest `finalized_checkpoint`
across _all_ branches (`store.finalized_checkpoint`), instead of the one
on the currently selected branch (`state.finalized_checkpoint`) when
deciding whether to ignore a block / blob because they are already
finalized.

etan-status committed:
Merge branch 'dev' into lc-gindexname

arnetheduck committed:
remove obsolete comment

hwwhww committed:
Remove extra tick

etan-status committed:
Gossip additional `LightClientFinalityUpdate` on new supermajority

When new finality is reached without supermajority sync committee
support, trigger another event push on beacon-API and libp2p once
the new finality gains supermajority support.

Without this, if the first `LightClientFinalityUpdate` that advances
finality has low participation, light clients monitoring gossip would
likely get stuck until the next time when finality advances (1 epoch).

Thanks to @DragonDev1906 for reporting this issue to Nimbus:

- https://github.com/status-im/nimbus-eth2/issues/5491

hwwhww committed:
Add `assert time >= store.time` to `on_tick`

tbenr committed:
Specify RPC byRoot blocks-sidecars elegibility

tbenr committed:
fix typo

tbenr committed:
Update specs/phase0/p2p-interface.md

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

tbenr committed:
Update specs/deneb/p2p-interface.md

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

etan-status committed:
Align gossip validation for aggregates with single attestations

A couple gossip validation rules are only specced out for single
un-aggregated attestations, but are also checked by implementations
for aggregates. This adds a copy of the missing gossip validation rules
to the aggregated attestation docs.

tbenr committed:
add failed on_block condition

tbenr committed:
rephrase

hwwhww committed:
Merge pull request #3548 from ethereum/fix-fc-script

Remove extra tick

tbenr committed:
Update specs/phase0/p2p-interface.md

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

tbenr committed:
apply suggestion

hwwhww committed:
Revert the spec. Only handle it in tests

hwwhww committed:
Merge branch 'dev' into on-tick-precondition

tbenr committed:
Update specs/deneb/p2p-interface.md

Co-authored-by: danny <dannyjryan@gmail.com>

tbenr committed:
Update specs/deneb/p2p-interface.md

Co-authored-by: danny <dannyjryan@gmail.com>

hwwhww committed:
cleanup leftover

tbenr committed:
remove the last consider

djrtwo committed:
Merge pull request #3550 from ethereum/on-tick-precondition

Check `assert time >= store.time` in fork-choice tests

tbenr committed:
from on_block to state_transition

hwwhww committed:
Add randomized block cases

tbenr committed:
simplify and add a new rule

djrtwo committed:
Merge pull request #3555 from ethereum/random-blob_kzg_commitment_merkle_proof

Add randomized block `blob_kzg_commitment_merkle_proof` cases

tbenr committed:
Update specs/phase0/p2p-interface.md

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

tbenr committed:
Update specs/deneb/p2p-interface.md

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

tbenr committed:
Update specs/deneb/p2p-interface.md

Co-authored-by: danny <dannyjryan@gmail.com>

tbenr committed:
remove gossip failure rules

hwwhww committed:
Move `blob_sidecar_{subnet_id}` to `Blob subnets` section

dankrad committed:
Add polynomial commitments for sampling

dankrad committed:
Add public method markers

djrtwo committed:
Apply suggestions from code review

hwwhww committed:
Merge pull request #3551 from tbenr/specify_rpc_by_root_elegibility

Specify when clients can serve block and sidecars in byRoot RPC methods

hwwhww committed:
bump version to v1.4.0-beta.5

djrtwo committed:
Merge pull request #3558 from ethereum/bump-version

bump version to v1.4.0-beta.5

djrtwo committed:
Merge pull request #3556 from ethereum/blob-subnets

Move `blob_sidecar_{subnet_id}` to `Blob subnets` section

dapplion committed:
fix tests for mainnet preset

hwwhww committed:
Merge pull request #3442 from dapplion/whisk-tests

Add WHISK fork only tests

hwwhww committed:
Apply suggestions from code review

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

hwwhww committed:
Minor fixes

hwwhww committed:
Extract the hardcoded spec names into a `specs.py` file

hwwhww committed:
Fix linter config

hwwhww committed:
Fix import

hwwhww committed:
update new-feature doc

hwwhww committed:
Use md_doc_paths.py `is_post_fork` in pyspec itself; clean up `create_genesis_state` helper

hwwhww committed:
Refactor `do_fork` with terrifying eval() and `PREVIOUS_FORK_OF`

hwwhww committed:
update doc

hwwhww committed:
update doc

hwwhww committed:
Add assertion to ensure eval usage is fine

dankrad committed:
SAMPLE -> CELL and cleanups

dankrad committed:
More sample -> cell renaming

dankrad committed:
Zero polynomial -> vanishing polynomial

dankrad committed:
Remove polynomial_coeff_to_eval

etan-status committed:
Use types for representing LC Merkle branches

For better legibility and alignment with `ethereum/beacon-APIs`, use
named types for the various Merkle branches used in the LC protocol.

- https://github.com/ethereum/beacon-APIs/blob/v2.4.2/types/altair/light_client.yaml#L2-L22

etan-status committed:
Capella addition

etan-status committed:
Align formatting

etan-status committed:
Avoid Mypy error during lint

etan-status committed:
Use proper types when dealing with LC Merkle proofs

hwwhww committed:
Merge pull request #3562 from ethereum/refactor-context

Refactor pyspec to reduce some hardcoded spec fork names

parithosh committed:
WIP: Update dockerisation (#3477)

* 🐳 add Dockerfile and dockerfile based script for consensus-spec tests

hwwhww committed:
Merge branch 'dev' into polynomial-commitments-sampling

hwwhww committed:
Add `KZG_SETUP_G1_MONOMIAL` to trusted setup files

hwwhww committed:
Fix some lint errors

hwwhww committed:
Add `PeerDAS` feature

hwwhww committed:
Fix fft helper and add basic test

hwwhww committed:
Add PeerDAS testing CI job

hwwhww committed:
Add `verify_cell_proof_batch` test

hwwhww committed:
Rename `recover_cells` to `recover_polynomial` and fix function signature. Add basic test case.

hwwhww committed:
Clean up Deneb specs. Add some type casting for using fft function

hwwhww committed:
Skip pylint check in CI

hwwhww committed:
Merge branch 'dev' into fork-choice-changes-for-confirmaton-rule

etan-status committed:
Merge branch 'dev' into lc-branchtypes

asn-d6 committed:
Randomize the test_recover_polynomial()

etan-status committed:
ensure `floorlog2` is available for use in `Custom types` section

etan-status committed:
Use `Custom types` section instead of `Helper types` for consistency

asn-d6 committed:
Work on hww's round-trip FFT test

etan-status committed:
Move `Vector` support from main section parser to types section parser

asn-d6 committed:
Fix linter

hwwhww committed:
Set `--bls-type` default value to `fastest` (#3567)

etan-status committed:
Extract LC migration test helpers to `light_client.py` module

The helpers needed to migrate LC data across forks are generic enough to
allow extraction from a specific test into the common `light_client.py`
helper module. This way, it is possible to use them from future tests.

Also complete the missing functions for `LightClientFinalityUpdate`,
`LightClientOptimisticUpdate` and `LightClientHeader` in same style.

etan-status committed:
Add guard when `fill_prev_epoch` is `True` in epoch `0`

`state_transition_with_full_block` takes two flags, `fill_cur_epoch`
and `fill_prev_epoch`, to control what attestations to include into
the new testing block.

While `fill_cur_epoch` uses `MIN_ATTESTATION_INCLUSION_DELAY` as the
minimum slot before taking effect, `fill_prev_epoch` results in an
annoying assert, requiring caller to work around the case of calling
the function as usual at the start.

This adds a similar minimum slot `SLOTS_PER_EPOCH` for `fill_prev_epoch`
before which the flag is ignored, same as for `fill_cur_epoch`.

etan-status committed:
Extract sync committee period <-> slot helpers to common LC test module

The two functions `compute_start_slot_at_sync_committee_period` and
`compute_start_slot_at_next_sync_committee_period` are currently only
available during LC `test_sync`. Move them to the common LC test module
so that they can be used from future tests as well.

etan-status committed:
Consider fork transitions when using `get_sync_aggregate` helper func

When calling `get_sync_aggregate` with a `signature_slot - 1` in a
future fork, i.e., the first slot of the new fork is missed, it uses
a wrong fork version for the signature. Fix that by correctly applying
fork transitions to the `signature_state`, if a fork schedule is given.

parithosh committed:
removing dockerfile ci test

parithosh committed:
update docker readme

hwwhww committed:
Update docker/README.md

hwwhww committed:
Merge pull request #3573 from parithosh/remove-docker-test

Removing docker CI test and add docker tutorial

hwwhww committed:
Merge pull request #3569 from etan-status/lc-migrationhelpers

Extract LC migration test helpers to `light_client.py` module

hwwhww committed:
Merge pull request #3571 from etan-status/lc-periodhelpers

Extract sync committee period <-> slot helpers to common LC test module

dankrad committed:
Add constant for multi cell verification challenge

dankrad committed:
Remove constants in favour of using compute_roots_of_unity

djrtwo committed:
Merge pull request #3506 from status-im/canonical-json-byte

ssz: `byte` type and canonical JSON mapping

dankrad committed:
Correct type info for verify_kzg_proof_multi_impl

dankrad committed:
Correct type definition for PolynomialCoeff

dankrad committed:
Doctoc

dankrad committed:
Linter

dankrad committed:
Forgot factor of 2 for roots_of_unity_extended

hwwhww committed:
Rename `peerdas` to `eip7594`

hwwhww committed:
Merge pull request #3557 from ethereum/polynomial-commitments-sampling

EIP-7594: Add cryptography specs for sampling

hwwhww committed:
Merge pull request #3537 from etan-status/lc-gindexname

Rename `_INDEX` generalized index constants to `_GINDEX`

hwwhww committed:
Merge pull request #3549 from etan-status/lc-finsend

Gossip additional `LightClientFinalityUpdate` on new supermajority

hwwhww committed:
Merge branch 'dev' into pr3563

etan-status committed:
Also bump `EXECUTION_PAYLOAD_GINDEX`

etan-status committed:
Merge branch 'dev' into lc-transitionacrossforks

hwwhww committed:
Merge pull request #3563 from etan-status/lc-branchtypes

Use types for representing LC Merkle branches

hwwhww committed:
Merge pull request #3572 from etan-status/lc-transitionacrossforks

Consider fork transitions when using `get_sync_aggregate` helper func

hwwhww committed:
Merge pull request #3570 from etan-status/ts-attprevepoch

Add guard when `fill_prev_epoch` is `True` in epoch `0`

asn-d6 committed:
Remove some trailing whitespace

asn-d6 committed:
Public methods must accept raw bytes

hwwhww committed:
Optimization: reduce `len()` call

asn-d6 committed:
Public methods: int -> uint64

djrtwo committed:
Merge pull request #3431 from saltiniroberto/fork-choice-changes-for-confirmaton-rule

Confirmation rule prerequisite - fork choice filter change

djrtwo committed:
Merge pull request #3552 from etan-status/gv-alignaggregate

Align gossip validation for aggregates with single attestations

djrtwo committed:
Merge pull request #3543 from etan-status/gv-finstore

Use fork choice to determine `finalized_checkpoint` in gossip validation

djrtwo committed:
Update VERSION.txt

CarlBeek committed:
Updates presets to use correct trusted setup from final ceremony output

hwwhww committed:
Merge pull request #3583 from ethereum/use_correct_trusted_setup

Use the correct `trusted_setup_4096.json`

hwwhww committed:
Merge pull request #3582 from ethereum/bump-version

bump verson.txt

asn-d6 committed:
Minor fixes based on review

asn-d6 committed:
Improve test_verify_cell_proof_batch()

hwwhww committed:
Add transition test case of non-empty pre-state `historical_roots`

hwwhww committed:
Apply suggestions from code review

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

hwwhww committed:
Merge pull request #3585 from ethereum/historical-roots-fork-transition

Add transition test case of non-empty pre-state `historical_roots`

hwwhww committed:
Merge pull request #3581 from ethereum/reduce-len-call

Optimization: reduce `len()` calls in `add_polynomialcoeff`

fradamt committed:
add changes from maxeb-min-viable-pyspec

asn-d6 committed:
Add helpers: recover_shifted_data() and recover_original_data()

fradamt committed:
typos

gurukamath committed:
add py.typed file mypy

When this is installed as a dependency in other projects, the py.typed file helps the other project to read the type hints from the consensus-specs

fradamt committed:
fixing things

fradamt committed:
exit tests working

fradamt committed:
basic deposit tests running

fradamt committed:
added some deposit + pending balance deposit tests

asn-d6 committed:
Merge pull request #3579 from asn-d6/peerdas_public_method_bytes

peerDAS: Public methods must accept raw bytes

fradamt committed:
process_execution_layer_withdraw_request does exit when amount == 0

fradamt committed:
adapted execution_layer_exit tests from eip7002

fradamt committed:
rename test files

fradamt committed:
Make initial slashing penalty negligible, as in https://github.com/michaelneuder/consensus-specs/pull/16

fradamt committed:
add consolidations based on https://github.com/michaelneuder/consensus-specs/pull/14 and fix presets/configs

fradamt committed:
fix tests

fradamt committed:
add consolidation_balance_to_consume and earliest_consolidation_epoch to state

fradamt committed:
fix small issues

fradamt committed:
add basic process_consolidation tests and helpers

fradamt committed:
only allow partial withdrawals from compounding creds, but full exits from all exec creds

fradamt committed:
do not allow consolidations with source == target (equivalent to exit)

fradamt committed:
add invalid process_consolidation tests

fradamt committed:
modified deposit tests from phase0

mkalinin commented on specs/_features/maxeb/beacon-chain.md:
````diff
@@ -0,0 +1,767 @@
+    return min(MAX_PER_EPOCH_ACTIVATION_EXIT_CHURN_LIMIT, get_churn_limit(state))
+```
+
+#### New `get_activation_exit_churn_limit`
````
```suggestion
#### New `get_consolidation_churn_limit`
```

mkalinin commented on specs/_features/maxeb/beacon-chain.md:
```diff
@@ -0,0 +1,767 @@
+    earliest_exit_epoch: Epoch  # Initialized with the max([v.exit_epoch for v in state.validators if v.exit_epoch != FAR_FUTURE_EPOCH]) + 1
+    consolidation_balance_to_consume: Gwei
+    earliest_consolidation_epoch: Epoch
+    pending_balance_deposits: List[PendingBalanceDeposit, 100000]
```
Considering 1 ETH as a minimal deposit, a number of deposits in the queue can’t exceed total amount of ETH, i.e. 2**27 would be a theoretical limit on this list. It is harder to get to a similar number with partial withdrawals but we can use the same limit for the beginning and probably restrict a minimal withdrawal amount which might make sense for other reasons — btw the limitation can be managed on CL side in the opposite to the deposits.

fradamt replied:
ah yeah sorry I don't remember why I put this limit at some point. For partial withdrawals why wouldn't we just put a much smaller limit? (enforced by the CL as you said, just noop when queue is full)

mkalinin commented on specs/_features/maxeb/beacon-chain.md:
```diff
@@ -0,0 +1,767 @@
+        state.consolidation_balance_to_consume -= consolidation_balance
+    else: # Consolidation doesn't fit in the current earliest epoch.
+        balance_to_process = consolidation_balance - state.consolidation_balance_to_consume
+        additional_epochs, remainder = divmod(balance_to_process, per_epoch_consolidation_churn)
```
This will error if the churn limit is `0`. I think we should explicitly process this case, probably in the `process_consolidation`, and also do not queue consolidation if it can’t be processed due to negligible churn limit

fradamt replied:
Good catch! Should we add `assert get_consolidation_churn_limit(state) > 0` at the top of `process_consolidation`?

mkalinin replied:
Yes, I think this is the only way we can enforce this constraint

mkalinin commented on specs/_features/maxeb/beacon-chain.md:
```diff
@@ -0,0 +1,767 @@
+    for_ops(body.deposits, process_deposit)
+    for_ops(body.voluntary_exits, process_voluntary_exit)
+    for_ops(body.bls_to_execution_changes, process_bls_to_execution_change) 
+    for_ops(body.execution_payload.withdraw_request, process_execution_layer_withdraw_request) # New
```
```suggestion
    for_ops(body.execution_payload.withdraw_requests, process_execution_layer_withdraw_request) # New
```

mkalinin commented on specs/_features/maxeb/beacon-chain.md:
```diff
@@ -0,0 +1,767 @@
+    withdrawals: List[Withdrawal] = []
+    consumed = 0
+    for withdrawal in state.pending_partial_withdrawals:
+        if withdrawal.withdrawable_epoch > epoch or len(withdrawals) == MAX_WITHDRAWALS_PER_PAYLOAD // 2:
```
I think we should add `MAX_PARTIAL_WITHDRAWALS_PER_PAYLOAD` and set it to some number, e.g. to a half of withdrawals limit as it is set in this line.

fradamt replied:
This seems a bit confusing because the sweep also does partial withdrawals. How about `MAX_PENDING_PARTIAL_WITHDRAWALS_PER_PAYLOAD`

fradamt committed:
Update specs/_features/maxeb/beacon-chain.md

fix function name in header

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

fradamt committed:
Update specs/_features/maxeb/beacon-chain.md

fix typo

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

dapplion commented on specs/_features/maxeb/beacon-chain.md:
```diff
@@ -493,19 +495,17 @@ def process_pending_balance_deposits(state: BeaconState) -> None:
-    active_balance_ceil = MIN_ACTIVATION_BALANCE if has_eth1_withdrawal_credential(validator) else MAX_EFFECTIVE_BALANCE
-    return min(state.balances[validator.index], active_balance_ceil)
+def get_active_balance(state: BeaconState, validator_index: ValidatorIndex) -> Gwei:
+    active_balance_ceil = MIN_ACTIVATION_BALANCE if has_eth1_withdrawal_credential(state.validators[validator_index]) else MAX_EFFECTIVE_BALANCE_MAXEB
```
Hsiao is gonna yell at you for this long lines

fradamt committed:
fix skipping pending consolidations when source is slashed

fradamt committed:
add basic process_pending_consolidations tests

fradamt committed:
rename maxeb to eip7251

mkalinin commented on specs/_features/eip7251/beacon-chain.md:
```diff
@@ -0,0 +1,763 @@
+
+    # New condition: only allow partial withdrawals witb compounding withdrawal credentials
+    is_full_exit_request = amount == 0
+    if not (is_full_exit_request or has_compounding_withdrawal_credential(validator)):
```
```suggestion
    if not (amount == FULL_EXIT_REQUEST_AMOUNT or has_compounding_withdrawal_credential(validator)):
```

Suggesting to introduce a constant for the amount denoting a full exit

fradamt replied:
would we ever want to set it to something other than 0?

fradamt committed:
add p2p-interface.md, update TOCs

fradamt committed:
apply suggestions from code review, small fixes

asn-d6 committed:
Small improvements on construct_vanishing_polynomial()

asn-d6 committed:
Initial docs added to recover_shifted_data() and recover_original_data()

asn-d6 committed:
Add helper method: construct_vanishing_polynomial()

hwwhww committed:
Apply suggestions from code review

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

asn-d6 committed:
Add defensive asserts in recover_polynomial()

asn-d6 committed:
New FIELD_ELEMENTS_PER_EXT_BLOB: number of cells in an extended blob

asn-d6 committed:
Update specs/_features/eip7594/polynomial-commitments-sampling.md

Co-authored-by: Hsiao-Wei Wang <hsiaowei.eth@gmail.com>

asn-d6 committed:
Update specs/_features/eip7594/polynomial-commitments-sampling.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

asn-d6 committed:
Update specs/_features/eip7594/polynomial-commitments-sampling.md

Co-authored-by: Justin Traglia <95511699+jtraglia@users.noreply.github.com>

hwwhww committed:
Merge pull request #3588 from gurukamath/mypy-types

add py.typed file for mypy

gurukamath committed:
correct the location of py.typed

hwwhww committed:
Merge pull request #3594 from gurukamath/mypy-types

correct `py.typed` location

etan-status committed:
Schedule Deneb for Mainnet

As discussed in https://github.com/ethereum/pm/issues/951, schedule
Deneb for epoch 269568.

hwwhww committed:
Set official mainnet `DENEB_FORK_EPOCH`

etan-status committed:
Merge commit '4878f1da30393582d67e2736a2c058430d05784f' into df-schedule

hwwhww committed:
Bump version to `1.4.0-beta.7`

hwwhww committed:
Merge pull request #3597 from etan-status/df-schedule

Schedule Deneb for Mainnet

hwwhww committed:
Merge pull request #3599 from ethereum/bump-version

Bump version to `1.4.0-beta.7`

hwwhww committed:
Update specs/_features/eip7594/polynomial-commitments-sampling.md

hwwhww committed:
Merge pull request #3591 from asn-d6/peerdas_refactor_recovery_polynomial

peerDAS: Initial refactor of recover_polynomial()

fradamt committed:
Merge branch 'ethereum:dev' into fradamt-maxeb-spec

dapplion commented:
Closing for
- https://github.com/ethereum/consensus-specs/pull/3618
